app/Jobs/ImportFileCsv.php

```diff
@@ -0,0 +1,244 @@
+
+        return $dataCsv;
+    }
+
+    /**
+     * Check the file title
+     *
+     * @param   array $firstLine
+     * @return bool
+     */
+    protected function checkHeaderCsv($firstLine)
+    {
+        $header = config('common.import.validation.file.header');
+        $map = array_map(function ($transHeader) {
+            return __($transHeader);
+        }, array_values($header));
+        
+        $firstLine = array_reverse($firstLine);     //always must reverse ???
+        $firstLineDecodeToUtf8 = $firstLine;
+        if (mb_check_encoding($firstLine, 'SJIS')) {
+            $firstLineDecodeToUtf8 = mb_convert_encoding($firstLine, "UTF-8", "SJIS");
+            
+        }
+
+        if ($firstLineDecodeToUtf8 !== $map) {
+            $message = __('import.message.error_header');
+            
+            throw new Exception($message);
+        }
+
+        return true;
+    }
+
+    /**
+     * Check the file content on each line
+     *
+     * @param   array $nextLine
+     * @param   array $totalColumn
+     * @param   int $row
+     * @return bool
+     */
+    protected function checkRowAndGetDataCsv($nextLine, $totalColumn, $row)
+    {
+        if (count($totalColumn) !== count($nextLine)) {
+            $message = __('import.message.error_element', ['row' => $row]);
+
+            throw new Exception($message);
+        }
+
+        if (mb_check_encoding($nextLine, 'SJIS')) {
+            $nextLineAndKey = array_combine(config('common.import.column'), array_reverse($nextLine));
+            $nextLineAndKey = mb_convert_encoding($nextLineAndKey, "UTF-8", "SJIS");
+        } else {
+            $nextLineAndKey = array_combine(array_keys(config('common.import.column')), $nextLine);
+        }
+
+        foreach ($nextLineAndKey as $key => $value) {
+            $column = (int)$key + 1;
+            
+            // check max length of data
+            if (mb_strlen($value) > config('common.import.validation.name.max')) {
+                $message = __('import.message.error_row_name', [
+                    'row' => $row,
+                    'column' => $column,
+                ]);
+
+                throw new Exception($message);
+            }
+
+            // check Age
+            if ($key == config('common.import.column.age')) {
+                if ( !(int)$value || (int)$value > (int)config('common.import.validation.age.max') || (int)$value < (int)config('common.import.validation.age.min') ) {
+                    $message = __('import.message.error_age', [
+                        'row' => $row,
+                        'column' => $column,
+                    ]);
+
+                    throw new Exception($message);
+                }
+            }
+
+            // check Unit exist or not
+            if ($key == config('common.import.column.unit_id')) {
+                $unitRepository = app(UnitRepositoryInterface::class);
+                $unit = $unitRepository->findByName($value);
```

Comment:
query trong vòng for. nếu file csv có 1000 rows thì sẽ mất 1000 câu query???

Reply:
oki anh, đoạn này hơi khó nhìn nên em ko phát hiện ra